### PR TITLE
Store account key in storage

### PR DESCRIPTION
**Pull request by cdloh**
No unit tests yet. But thoughts?

I've left it so that if no filepath is supplied it defaults to storage but if one is supplied it'll try to load it off the FS. Means that present configs wont break.

### CODE REVIEW

**Comment by fffonion** [comment]
Hi @cdloh, I like the idea, but one major thought: since `AUTOSSL.init` is supposed to be called in the init_by_lua_* directive where cosocket is not available, so we can't actually use other storage except from file or shm. One possible workaround I can think of is to lazy load the account key, in init_worker phase.

**Comment by cdloh** [comment]
Ah I'd forgotten cosockets aren't available there! Moving it into the init_worker phase would mean changing the way the acme client is loaded as well as it tries to get the account thumbprint. Would you want to move creating the client into the init_worker as well or change the way that the fingerprint in checked?

If it was moved into the init_worker phase would you want to put a lock around it so only one worker tries to load the key?

**Comment by fffonion** [comment]
I think it's fine to move to init_worker, but let's not make it a breaking change. Probably we can allow to not specify the account_key during init, but invent a new function to set it afterwards.

> would you want to put a lock around it so only one worker tries to load the key?

We can make it simpler by just running it on worker 0

**Comment by cdloh** [comment]
Hmm would that require moving the self references in client to a shared dict?

I've got it working but the way it's loading it's only effecting worker 0. So the rest of the workers dont know about the account key.

Apologies this is the first time I've stepped this deep in resty modules.

**Comment by fffonion** [comment]
@cdloh ah i think you're right... since it'll be a copy on write. Let's still load the key from storage in every worker then.
Then it comes to another question, how do we only generate the account key once if it's unset... I tried to sort out the flow a bit:
- If account_key_path set, read the file; if not exist, bail out.
- If account_key_path is nil, always generate a new key in init_by_lua phase; in init_worker phase, try load from storage; if not found, use the previous generated key, and store in storage.
This might make the nginx start time unnecessary longer, but reload are not affected. How do you think?

**Comment by cdloh** [comment]
That sounds great. I've implemented my understanding of that if you want to take a look?

File: lib/resty/acme/autossl.lua
Comment (fffonion) [inline lib/resty/acme/autossl.lua:456]: this part of logic seems weird and `generated_key` seems unset

File: lib/resty/acme/client.lua
Comment (fffonion) [inline lib/resty/acme/client.lua:96]: let's make this function a local and set in _M, so we can reuse in new(), also maybe call it set_account_key?

```lua
local function set_account_key(self, account_key)
....
end

function _M.new(self)
...
if conf.account_key then
    local ok, err = set_account_key(self, conf.account_key)
    if err then
         return nil, errr
    end
...
end


_M.set_account_key = set_account_key
```

File: lib/resty/acme/autossl.lua
Comment (fffonion) [inline lib/resty/acme/autossl.lua:480]: this can be omitted

File: lib/resty/acme/autossl.lua
Comment (fffonion) [inline lib/resty/acme/autossl.lua:73]: let's add trailing `:` to the prefix to be consistent with others

File: lib/resty/acme/autossl.lua
Comment (fffonion) [inline lib/resty/acme/autossl.lua:468]: let's not short circuit it inside the function, instead only do:
```lua
if filepath then
   AUTOSSL.load_account_key(filepath)
end
```

File: lib/resty/acme/autossl.lua
Comment (cdloh) [reply lib/resty/acme/autossl.lua:456]: Whops good catch.

File: lib/resty/acme/autossl.lua
Comment (fffonion) [inline lib/resty/acme/autossl.lua:361]: with `:` you don't need to pass the self reference into function, we can just 
`AUTOSSL.client:set_account_key(account_key)`

File: lib/resty/acme/client.lua
Comment (fffonion) [inline lib/resty/acme/client.lua:52]: let's use a `return_value, error_msg` pattern here (also we need to update from the caller side)
with this function it will be `return false, <err>` on error and `return true` on success.

File: lib/resty/acme/autossl.lua
Comment (fffonion) [inline lib/resty/acme/autossl.lua:452]: let's bubble up errors and only `error()` out from the up most function

File: lib/resty/acme/autossl.lua
Comment (fffonion) [inline lib/resty/acme/autossl.lua:291]: let's add a comment describing why we want to always generate it here

**Review by fffonion — COMMENTED** [review]
LGTM overall, some style+comment. The CI for PRs are broken, I'll rerun the test after it got merged.

**Review by fffonion — APPROVED** [review]

**Comment by fffonion** [comment]
Thanks @cdloh! I'll merge this when CI is fixed, currently travis blocks public repository build and I'm trying to migrate to github actions 🤦 .

**Comment by fffonion** [comment]
I still didn't figure out how to make github actions happy. Will just go ahead and merge for now.

**Comment by cdloh** [comment]
@fffonion want a hand with that? I'm happy to attempt an MR.

**Comment by fffonion** [comment]
@cdloh I have an WIP branch at https://github.com/fffonion/lua-resty-acme/tree/github-actions, if you are familiar with github actions I could use a hand to setup the consul and vault service container : )

**Pull request merged**
